module/os/linux/zfs/zfs_file_os.c

```diff
@@ -0,0 +1,312 @@
+#include <sys/stat.h>
+#include <sys/file.h>
+
+#include <linux/falloc.h>
+#include <linux/fs.h>
+#include <linux/uaccess.h>
+#ifdef HAVE_FDTABLE_HEADER
+#include <linux/fdtable.h>
+#endif
+/*
+ * zfs_file_open -> filp_open
+ * zfs_file_close -> filp_close
+ * zfs_file_seek -> vfs_llseek
+ * zfs_file_sync -> spl_filp_fsync
+ * zfs_file_pwrite -> spl_kernel_write
+ * zfs_file_pread -> spl_kernel_read
+ * zfs_file_stat -> vfs_getattr
+ * zfs_file_unlink -> vfs_unlink
+ * zfs_file_get -> fget
+ * zfs_file_put -> fput
+ */
+
+int
+zfs_file_open(const char *path, int flags, int mode, zfs_file_t **fpp)
+{
+	struct file *filp;
+	int saved_umask;
+
+	ASSERT(flags & (FWRITE | FREAD));
+
+	if (!(flags & FCREAT) && (flags & FWRITE))
+		flags |= FEXCL;
+
+	/*
+	 * Note for filp_open() the two low bits must be remapped to mean:
+	 * 01 - read-only  -> 00 read-only
+	 * 10 - write-only -> 01 write-only
+	 * 11 - read-write -> 10 read-write
+	 */
+	flags--;
+
+	if (flags & FCREAT)
+		saved_umask = xchg(&current->fs->umask, 0);
+
+	filp = filp_open(path, flags, mode);
+
+	if (flags & FCREAT)
+		(void) xchg(&current->fs->umask, saved_umask);
+
+	if (IS_ERR(filp))
+		return (-PTR_ERR(filp));
+
+	*fpp = filp;
+	return (0);
+}
+
+void
+zfs_file_close(zfs_file_t *fp)
+{
+	filp_close(fp, 0);
+}
+
+int
+zfs_file_getattr(zfs_file_t *filp, zfs_file_attr_t *zfap)
+{
+	struct kstat stat;
+	int rc;
+
+	ASSERT(filp);
+	ASSERT(zfap);
+
+#if defined(HAVE_4ARGS_VFS_GETATTR)
+	rc = vfs_getattr(&filp->f_path, &stat, STATX_BASIC_STATS,
+	    AT_STATX_SYNC_AS_STAT);
+#elif defined(HAVE_2ARGS_VFS_GETATTR)
+	rc = vfs_getattr(&filp->f_path, &stat);
+#else
+	rc = vfs_getattr(filp->f_path.mnt, filp->f_dentry, &stat);
+#endif
+	if (rc)
+		return (-rc);
+
+
+	zfap->zfa_size = stat.size;
+	zfap->zfa_mode = stat.mode;
+
+	return (0);
+}
+
+int
+zfs_file_fsync(zfs_file_t *filp, int flags)
+{
+	int datasync = 0;
+	int error;
+	int fstrans;
+
+	ASSERT(filp);
+
+	if (flags & FDSYNC)
+		datasync = 1;
+
+	/*
+	 * May enter XFS which generates a warning when PF_FSTRANS is set.
+	 * To avoid this the flag is cleared over vfs_sync() and then reset.
+	 */
+	fstrans = __spl_pf_fstrans_check();
+	if (fstrans)
+		current->flags &= ~(__SPL_PF_FSTRANS);
+	error = -vfs_fsync(filp, datasync);
+	if (fstrans)
+		current->flags |= __SPL_PF_FSTRANS;
+
+	return (error);
+}
+
+int
+zfs_file_pwrite(zfs_file_t *fp, const void *buf, size_t count, loff_t *offp,
+    ssize_t *resid)
+{
+	ssize_t rc;
+
+#if defined(HAVE_KERNEL_READ_PPOS)
+	rc = kernel_write(fp, buf, count, offp);
+#else
+	mm_segment_t saved_fs;
+
+	saved_fs = get_fs();
+	set_fs(KERNEL_DS);
+
+	rc = vfs_write(fp, (__force const char __user __user *)buf,
+	    count, offp);
+
+	set_fs(saved_fs);
+#endif
+	if (rc < 0)
+		return (-rc);
+	if (resid)
+		*resid = count - rc;
+	return (0);
+}
+
+int
+zfs_file_write(zfs_file_t *fp, const void *buf, size_t count, ssize_t *resid,
+    int flags)
+{
+	zfs_file_attr_t zfa;
+	int rc;
+
+	if (flags & FAPPEND) {
```

Comment:
Similar to user space, I didn't see any callers which pass `FAPPEND`.  If I missed it please point me to it!